SpatialAveragePooling.cu

```diff
@@ -241,78 +181,91 @@ static int cunn_SpatialAveragePooling_updateGradInput(lua_State *L)
-
-    // cuda blocks & threads:
-    int yblocks = (int)(16L / nInputPlane);
-    yblocks = yblocks < 1 ? 1 : yblocks;
-    dim3 blocks(nInputPlane,yblocks);
-    dim3 threads(32,8);
-
-    // run updateGradInput kernel
-    if (kH == dH && kW == dW) {
-      subgradinput <<<blocks, threads, 0, THCState_getCurrentStream(state)>>> (gradInput_data, gradOutput_data,
-                                          nInputPlane, nInputRows, nInputCols,
-                                          kH, kW, dH, dW);
-    } else {
-      subgradinputAtomic <<<blocks, threads, 0, THCState_getCurrentStream(state)>>> (gradInput_data, gradOutput_data,
-                                                nInputPlane, nInputRows, nInputCols,
-                                                kH, kW, dH, dW);
-    }
-  } else {
-    long nInputCols = input->size[3];
-    long nInputRows = input->size[2];
-    long nInputPlane = input->size[1];
-    long nbatch = input->size[0];
-
-    float *gradOutput_data = THCudaTensor_data(state, gradOutput);
-    float *gradInput_data;
-
-    THCudaTensor_resizeAs(state, gradInput, input);
-    THCudaTensor_zero(state, gradInput);
-    gradInput_data = THCudaTensor_data(state, gradInput);
-
-    // cuda blocks & threads:
-    int yblocks = (int)(16L / nInputPlane);
-    yblocks = yblocks < 1 ? 1 : yblocks;
-    dim3 blocks(nInputPlane*nbatch,yblocks);
-    dim3 threads(32,8);
-
-    // run updateGradInput kernel
-    if (kH == dH && kW == dW) {
-      subgradinput <<<blocks, threads, 0, THCState_getCurrentStream(state)>>> (gradInput_data, gradOutput_data,
-                                          nInputPlane, nInputRows, nInputCols,
-                                          kH, kW, dH, dW);
-    } else {
-      subgradinputAtomic <<<blocks, threads, 0, THCState_getCurrentStream(state)>>> (gradInput_data, gradOutput_data,
-                                                nInputPlane, nInputRows, nInputCols,
-                                                kH, kW, dH, dW);
-    }
+    nInputCols = input->size[2];
+    nInputRows = input->size[1];
+    nInputPlane = input->size[0];
+    batchSize = 1;
+  }
+  else
+  {
+    nInputCols = input->size[3];
+    nInputRows = input->size[2];
+    nInputPlane = input->size[1];
+    batchSize = input->size[0];
   }
 
+  if(ceil_mode) {
+    nOutputCols = ceil(float(nInputCols - kW + 2*padW) / float(dW)) + 1;
+    nOutputRows = ceil(float(nInputRows - kH + 2*padH) / float(dH)) + 1;
+  }
+  else {
+    nOutputCols = floor(float(nInputCols - kW + 2*padW) / float(dW)) + 1;
+    nOutputRows = floor(float(nInputRows - kH + 2*padH) / float(dH)) + 1;
+  }
+  if (padW || padH)
+  {
+    // ensure that the last pooling starts inside the image
+    // needed to avoid problems in ceil mode
+    if ((nOutputRows - 1)*dH >= nInputRows + padH)
+      --nOutputRows;
+    if ((nOutputCols  - 1)*dW >= nInputCols  + padW)
+      --nOutputCols;
+  }
+
+  gradOutput = THCudaTensor_newContiguous(state, gradOutput);
```

Comment:
Redundant `new_Contiguous` call, has already been called in https://github.com/torch/cunn/pull/134/files#diff-96a3e9b178afa9682543a117c11f623aR194